wincent/liferay-portal
# LPS-110927 Update Liferay.Portlet.openWindow utility

Hey @wincent ,

~~This PR contains LPS-110926 commits, so we cannot forward it. We can safely assume https://github.com/brianchandotcom/liferay-portal/pull/87801 will be merged with no additional changes, so we can proceed with review.~~

In this PR:
- We are adding `Liferay.Portlet.openModal` utility. It is implemented as a wrapper around Clay-based `Liferay.Util.openModal`. It is a direct replacement of AUI-based `Liferay.Portlet.openWindow`.
- We are extending `openModal` API to support `headerHTML`.
- We are implementing support for form buttons within iframe
- We are changing implementation of `Liferay.Portlet.openWindow`, to be a thin wrapper around `Liferay.Portlet.openModal`
- We are replacing all usages of `Liferay.Portlet.openWindow`

To test staging-configuration-web
1. Enable Local Live staging (Control Panel > Site > Publishing > Staging)
1. Go to home page > Staging tab
1. Open home page > Hello World porlet options > Staging

To test export-import-web
1. Open home page > Hello World porlet options > Export / Import

To test notifications-web
1. Click user icon on top right > Notifications
1. Open Notifications porlet options > Configuration

To test portal-web
1. Open home page > Hello World porlet options > Configuration

Notes:
- We could have just changed implementation of `Liferay.Portlet.openWindow`, but I like to have consistent `*.openModal` for Clay-based utils.
- We could have just marked `Liferay.Portlet.openWindow` as deprecated, like we did with `SignInModal`, but we want to aggressively remove eager AUI code, like in `portlet.js`. Among other things, it is a performance bottleneck for AUI preloading and requires special babel configuration.
- We need to implement support for buttons withing iframe, see the reasoning [here](https://github.com/liferay-echo/liferay-portal/pull/237#issuecomment-614559365). While most buttons work by default, we need to have special handling for 'cancel' button. It needs to close modal. To achive this, I added delegation, defined on iframe document node. To be able to do this, I needed iframe `ref`. As `ClayModal.Body` does not expose this, I replaced it with plain `<div>`. All of this causes a bit of code duplication, with CSS classes around body. 
Is there a batter way to this this? Could `ClayModal` expose iframe `ref`, or is that a bad practice? (@kresimircoko this might be interesting for you).
- On the track of https://github.com/wincent/liferay-portal/pull/208#discussion_r411295441, I tweaked body html to implement `url` / `bodyHTML` exclusion, each leading to it's own subcomponent.

Please review the code.

Thank you!

/cc @jbalsas
